Full rework of the TF input/output embeddings and bias resizing

jplu opened the pull request:
# What does this PR do?

This PR 100% reworks the entire process of input/output and bias resizing. Now the exceptions are better handled including the names that now are always similar. The corresponding tests have also been entirely reworked and now have a better coverage of this feature.

This PR adds a small breaking change. Now the `get_input_embeddings` methods returns the weights and not anymore the embedding layer.

LysandreJik commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -584,44 +584,55 @@ def __init__(self, config, *inputs, **kwargs):
         self.name_or_path = config.name_or_path
 
-    def get_input_embeddings(self) -> tf.keras.layers.Layer:
+    def get_input_embeddings(self) -> tf.Variable:
```
And the `get_output_embeddings` method still returns a `tf.keras.layers.Layer`

jplu replied:
They all return a `tf.Variable` that are the embeddings/bias weights. This is now necessary in order to be sure to have the proper attribute and not guessing if it is `word_embeddings`, `weights`, `bias`, `decoder_bias`, etc... and to make this as generic as possible, including for the naming.

jplu replied:
By "naming" I mean that now, we don't need anymore to manually build the name of each resized weights as thanks to these changes, they are always the same.

jplu replied:
`get_input_embeddings` still returns a layer since the last commit 👍

sgugger commented:
I haven't reviewed in detail yet, but just looking at the API with the number of things to change for ALBERT (and in terms of line of code) is a hard pass for me. Overriding the resize method as was done before was way easier, this adds too much complexity.

jplu commented:
I understand that it is a big update. Nevertheless, the way it was done before didn't worked and was quite buggy (the tests basically was testing almost nothing) and to make the resizing properly working, these changes are necessary.

jplu commented:
In all cases I'm open to any suggestion that will reduce the number of changes :)

jplu commented:
@sgugger I tried a new approach for the resizing that reduce a lot the changes in each model implementation. I have done my test only on ALBERT for now, can you recheck it and let me know what you think about it.

patrickvonplaten commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -584,59 +614,121 @@ def __init__(self, config, *inputs, **kwargs):
+
+        if hasattr(self, "get_lm_head"):
+            lm_head = self.get_lm_head()
+            try:
```
not a huge fan of this function. It can silently suppress errors and result in some unexpected behavior IMO. I really prefer not to use try....except statements in such a function

patrickvonplaten replied:
Why do we need this in the first place?

jplu replied:
Because we need to be sure that the model is entirely built and not partially.

patrickvonplaten commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -584,59 +614,121 @@ def __init__(self, config, *inputs, **kwargs):
+            try:
+                return main_layer.get_input_embeddings()
+            except AttributeError:
+                self(self.dummy_inputs)
```
What is the use case of this except statement -> can you clarify a bit why we need this functionality?

jplu replied:
Because we need to be sure that the model is entirely built and not partially.

patrickvonplaten reviewed (commented):
@jplu - can you clarify a bit why we need this new "TF input/output embeddings resizing" PR and why the previous PR: https://github.com/huggingface/transformers/pull/8657 was unsatisfactory?

jplu commented:
Ok I will clarify this a bit more:

1. The fist most important issue in the current implementation is that the resizing is not graph compilation+ execution compliant because of the usage of `numpy` and `tensor.numpy()` calls and then not usable in such cases.
2. The naming was depending of where the build was coming from, that's why we needed the `get_prefix_bias_name` for the bias  and the manual build of the embeddings names. Which was a temporary fix, because it is very error prone, because the naming depends of several other things that are not taken into account into this manual build.
3. Resizing was not working for some models, such as BART which doesn't work properly and raises an error. (Proof that the tests was not testing almost nothing)
4. The current resizing has two issues when resizing when we instantiate a model from scratch: either it raises an attribute error because the model is not fully built (weights not instantiated) and get a wrong naming, and then if we save the model with this wrong naming we cannot save/load it properly because the naming doesn't correspond to the current architecture.
5. All the weights names across the models don't share the same names sometimes `embeddings.word_embeddings`, sometimes `shared.weight`, sometimes `lm_head.bias`, sometimes `lm_loss.bias`, sometimes `mlm.bias`,  sometimes `mlm.predictions.bias` and many other ways...

As stated in #8657 this was just a temporary fix to go to the quickest way to wait for the real rework.

This PR aims to solve all these issues and bring something more generic and less error prone.

sgugger commented:
I personally never understood that #8657 was a quick fix that was needing another PR afterwards. We cannot operate by adding new methods in one release then breaking them or deleting them in the next so the works that was done in #8657 needs to be built upon not destroyed (and please, say in bold next time you are just making a quick fix as I would never have approved #8657 to be merged had I known...)

So before we review this, the following need to be addressed:
- `get_input_embeddings` needs to keep the same return type
- `get_output_embeddings` needs to keep the same return type
- `get_output_layer_with_bias` can't disappear
- `get_prefix_bias_name` can't disappear

This is annoying but this is why we usually don't merge a half-baked fix introducing new APIs, we can't break that after.

jplu commented:
We can keep this for the next major release.

What you ask is doable but will make the codebase more complicated. I will rework this.

jplu commented:
I have just done the following restore:

- `get_input_embeddings` still returns a layer
- `get_output_embeddings` still returns a layer
- `get_output_layer_with_bias` is back
- `get_prefix_bias_name` is back

The old and new approach was much more compliant than I thought so it was easier to restore what @sgugger asked. Really sorry for the misunderstanding, I will clearer next time.

sgugger reviewed (commented):
It's better without the breaking changes, thanks for adapting.

Reviewing this, I see a lot of:
```
   def get_output_embeddings(self):
        return self.input_embeddings

    def set_output_embeddings(self, value):
        self.input_embeddings.weight = value
        self.input_embeddings.vocab_size = shape_list(value)[0]

    def get_bias(self):
        return {"bias": self.bias}

    def set_bias(self, value):
        self.bias = value["bias"]
        self.vocab_size = shape_list(value["bias"])[0]
```
which makes me think there is a better way to code the default in `modeling_tf_utils`.

At the same time, I have a bad feeling this is currently breaking the weight-tying which is currently untested (cf my comment in `_resize_token_embeddings`, L803) so I think we should first have a test of the weight-tying before merging.

jplu commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -656,143 +757,178 @@ def resize_token_embeddings(self, new_num_tokens=None) -> tf.Variable:
+            self.set_bias(new_lm_head_bias)
+
+        # if word embeddings are not tied, make sure that lm head decoder is resized as well
+        if self.get_output_embeddings() is not None:
```
If you continue to follow the logic in this `if`, you can see that in the `_get_resized_lm_head_decoder()` there is a test if `input_embeddings` and `output_embeddings` are equals with
```
if old_lm_head_decoder is not None and not is_input_output_equals
```

So yes, I confirm that the tying is well tested 👍

jplu commented:
@sgugger I should have addressed all your comments :)

> which makes me think there is a better way to code the default in modeling_tf_utils.

Share your thoughts 😉

LysandreJik commented on tests/test_modeling_tf_dpr.py:
```diff
@@ -209,6 +209,9 @@ def test_dpr_reader_model(self):
         config_and_inputs = self.model_tester.prepare_config_and_inputs()
         self.model_tester.create_and_check_dpr_reader(*config_and_inputs)
 
+    def test_model_common_attributes(self):
+        pass
+
```
Is this something to fix later on?

jplu replied:
Oops! Making an update for this one!

jplu replied:
Done!

LysandreJik commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -656,143 +769,177 @@ def resize_token_embeddings(self, new_num_tokens=None) -> tf.Variable:
+
+        return model_embeds
+
+    def _find_weights(self, embedding_layer):
+        if hasattr(embedding_layer, "word_embeddings"):
+            return embedding_layer.word_embeddings
+        elif hasattr(embedding_layer, "weight"):
+            return embedding_layer.weight
+        elif hasattr(embedding_layer, "decoder"):
+            return embedding_layer.decoder
```
We should keep this method in mind when defining new LM head layers otherwise this would grow to be unmaintainable. It's unfortunate that we need to have such a mapping here, but I understand why it's necessary cf your comment [here](https://github.com/huggingface/transformers/pull/9193#discussion_r546039935).

jplu replied:
I'm currently rethinking the way we implement de embeddings and a part of this "rethinking" is how to get rid of these attribute detection, or at least minimize the side effects.

patrickvonplaten replied:
Can we change the naming a bit maybe? just `_find_weights` is a bit too generic IMO => `_get_word_embedding_weight` ? Also (not sure here) maybe raise if nothing is found instead of silent None?

LysandreJik reviewed (approved):
This looks better than the previous version, the tests are extensive enough. I find it a bit weird that `get_bias` returns a dict, but I understand why that's necessary.

Also it feels like a big bunch of the `test_model_common_attributes` could be refactored in the `test_modeling_tf_common.py`, but it's okay to leave it like this right now as it's explicit enough.

I may be missing something, but as Sylvain said, I'm not seeing any tests related to the tying itself.

jplu commented:
Thanks @LysandreJik! For the tying test look in the `_get_resized_lm_head_decoder()` method.

LysandreJik commented:
I mean I'm not seeing a test that checks `get_input_embeddings() == get_output_embeddings()` when weights are tied, but I may be missing something here. 

I know these two generally point to the same tensors, but no always, do they?

jplu commented:
> I mean I'm not seeing a test that checks get_input_embeddings() == get_output_embeddings() when weights are tied, but I may be missing something here.

Yes, there is a test for this, line 884 in `modeling_tf_utils`.

> I know these two generally point to the same tensors, but no always, do they?

Yes they always point to the same tensor when they equals, 100% sure.

patrickvonplaten commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -586,58 +626,131 @@ def __init__(self, config, *inputs, **kwargs):
+        if value is not None:
+            main_layer = getattr(self, self.base_model_prefix, self)
+
+            if main_layer is not self:
```
I'd slighly rewrite this as follows

```python

main_layer = getattr(self, self.base_model_prefix)

if main_layer is None:
    raise NotImplementedError(<here a good error message>)

try: 
     ...
except
    ...
```

patrickvonplaten commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -586,58 +626,131 @@ def __init__(self, config, *inputs, **kwargs):
+            lm_head = self.get_lm_head()
+            try:
+                lm_head.set_output_embeddings(value)
+            except AttributeError:
```
a comment here would be nice, maybe even a logger.info

patrickvonplaten commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -586,58 +626,131 @@ def __init__(self, config, *inputs, **kwargs):
             :obj:`tf.keras.layers.Layer`: The layer that handles the bias, None if not an LM model.
         """
-        return None
+        warnings.warn(
```
Nice!

patrickvonplaten commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -586,58 +626,131 @@ def __init__(self, config, *inputs, **kwargs):
+                self(self.dummy_inputs)
+                lm_head.set_bias(value)
+
+    def get_lm_head(self) -> tf.keras.layers.Layer:
```
maybe also a small comment that models that have a  lm head should overwrite this method.

patrickvonplaten commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -656,143 +769,177 @@ def resize_token_embeddings(self, new_num_tokens=None) -> tf.Variable:
+        new_lm_head_bias = {}
+
+        for attr, weight in old_lm_head_bias.items():
+            first_dim, old_num_tokens = (None, shape_list(weight)[0]) if tf.rank(weight) == 1 else shape_list(weight)
```
`first_dim` is always `hidden_size_dim` no? Think `hidden_size_dim` is a better name

jplu replied:
No, not always, in BART the shape of `final_logits_bias` is `[1, num_tokens]` and I don't think that 1 represents the `hidden_size_dim`.

patrickvonplaten commented on src/transformers/models/t5/modeling_tf_t5.py:
```diff
@@ -1191,24 +1189,24 @@ def __init__(self, config, *inputs, **kwargs):
+            return tf.transpose(self.lm_head.kernel)
+
+    def set_output_embeddings(self, value):
+        if value is not None:
```
why would the value be `None`? I'd delete this if

jplu replied:
This is a public method, we never know how this method will be used, so to prevent at least this misusage I have added this condition.

patrickvonplaten replied:
It's better to delete it still. If a user calls 

```python

t5.set_output_embeddings(self, word_embeddings)
```

and `word_embeddings` is None for whatever reason this should raise an error IMO not just silently ignore it. If we ignore it the user might have a much harder time finding the error.

jplu replied:
Ok, let's remove this condition then.

patrickvonplaten commented on src/transformers/models/transfo_xl/modeling_tf_transfo_xl.py:
```diff
@@ -925,6 +903,9 @@ def __init__(self, config):
         )
 
+    def _resize_token_embeddings(self, new_num_tokens):
+        return None
```
raise NotImplemented() ? This looks like a silent bug to me atm

patrickvonplaten commented on templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_tf_{{cookiecutter.lowercase_modelname}}.py:
```diff
@@ -460,6 +460,20 @@ def build(self, input_shape):
         self.bias = self.add_weight(shape=(self.vocab_size,), initializer="zeros", trainable=True, name="bias")
```
Could you update the Seq2Seq Templates as well? => this would be veeery helpful for the TF Bart split PR

patrickvonplaten replied:
And I'm afraid the `test_modeling_tf_{{cok...` as well

patrickvonplaten commented on src/transformers/models/bart/modeling_tf_bart.py:
```diff
@@ -1134,22 +1158,20 @@ def __init__(self, config, *inputs, **kwargs):
+    def get_output_embeddings(self):
+        return self.get_input_embeddings()
+
+    def set_output_embeddings(self, value):
```
We could also do the check here with `if config.tie_word_embeddings` like in TF no?

patrickvonplaten replied:
Bart should provide this functionality out of the box I think

jplu replied:
Can you handle this in your BART refactor?

patrickvonplaten replied:
never mind was just a (nit) - think it's not too important :-)

patrickvonplaten reviewed (commented):
Cool, looks good to me in general! 

I left a couple of comments. It would be great if you can fully update both the `modeling_tf_...` and the `test_modeling_tf_...` templates as well :-)

jplu commented:
I should have addressed all the comments.

patrickvonplaten commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -656,143 +770,177 @@ def resize_token_embeddings(self, new_num_tokens=None) -> tf.Variable:
+
+        return model_embeds
+
+    def _find_weights(self, embedding_layer):
```
still my comment here on the function naming -> thin `_find_weights` is not ideal.

-> a) It's only used for word embeddings
-> b) I think it's not a "find" method but just a "get" method. We don't have any methods called "find" in Transformers as far as I know

-> maybe rename it to `_get_word_embedding_weight`?

jplu replied:
Ooops thanks for the notice!!

jplu replied:
Ok I changed the name everywhere!

patrickvonplaten reviewed (approved):
Looks good to me now - thanks for taking care of this and also adding it to the model templates!

patrickvonplaten commented:
Ah yeah, we probably need a rebase here since TF-serving just got merged :-/

jplu commented:
Arf good point!

jplu commented:
@patrickvonplaten the test `TFLEDModelTest::test_pt_tf_model_equivalence` seems very flaky, it looks like that it randomly pass/fail.

jplu commented:
Good to merge for me now :)

patrickvonplaten commented:
> @patrickvonplaten the test `TFLEDModelTest::test_pt_tf_model_equivalence` seems very flaky, it looks like that it randomly pass/fail.

Just fixed it: https://github.com/huggingface/transformers/pull/9459

jplu commented:
@sgugger any objection to merge this PR?

sgugger reviewed (approved):
Looks good to me!
